tao_compiler/mlir/disc/transforms/lower_tf.cc

```diff
@@ -628,12 +633,438 @@ LogicalResult ConvertDequantizeOp::dequantize(TF::DequantizeOp op,
+                                              Builder* builder) {
+  RankedTensorType ty = RankedTensorType::get(
+      {static_cast<int64_t>(values.size())}, builder->getIntegerType(64));
+  return mlir::DenseIntElementsAttr::get(ty, values);
+}
+
+NamedAttribute getConvDimensionNumbersAttr(
+    ArrayRef<int64_t> spatial_dims, tensorflow::TensorFormat format,
+    Builder* builder,
+    tensorflow::FilterTensorFormat filter_format = tensorflow::FORMAT_HWIO) {
+  int64_t num_spatial_dims = spatial_dims.size();
+  int64_t num_dims = num_spatial_dims + 2;
+
+  int64_t batch_dim = GetTensorBatchDimIndex(num_dims, format);
+  int64_t feature_dim = GetTensorFeatureDimIndex(num_dims, format);
+
+  int64_t kernel_input_feature_dim = num_spatial_dims;
+  int64_t kernel_output_feature_dim = num_spatial_dims + 1;
+  SmallVector<int64_t, 4> kernel_spatial_dimensions;
+  kernel_spatial_dimensions.resize(num_spatial_dims);
+  if (filter_format == tensorflow::FORMAT_OHWI) {
+    kernel_input_feature_dim = num_spatial_dims + 1;
+    kernel_output_feature_dim = 0;
+    std::iota(kernel_spatial_dimensions.begin(),
+              kernel_spatial_dimensions.end(), 1);
+  } else {
+    assert(filter_format == tensorflow::FORMAT_HWIO);
+    // Filters data_format is always HWIO so input channels dimension is after
+    // all spatial dimensions.
+    std::iota(kernel_spatial_dimensions.begin(),
+              kernel_spatial_dimensions.end(), 0);
+  }
+
+  return builder->getNamedAttr(
+      "dimension_numbers",
+      mhlo::ConvDimensionNumbersAttr::get(
+          builder->getContext(), batch_dim, feature_dim, spatial_dims,
+          kernel_input_feature_dim, kernel_output_feature_dim,
+          kernel_spatial_dimensions, batch_dim, feature_dim, spatial_dims));
+}
+
+LogicalResult parseConvParams(ConvParams& params, OpBuilder& b, Location& loc) {
+  Operation* op = params.op;
+  auto input_ty = params.input.getType().dyn_cast<RankedTensorType>();
+  auto filter_ty = params.filter.getType().dyn_cast<RankedTensorType>();
+  int num_dims = input_ty.getRank();
+  int64_t num_spatial_dims = num_dims - 2;
+  Type shape_scalar_type = b.getIntegerType(32);
+
+  tensorflow::TensorFormat data_format = tensorflow::TensorFormat::FORMAT_NHWC;
+  tensorflow::Padding padding;
+  auto paddingStr = op->getAttrOfType<StringAttr>("padding").str();
+  if (!tensorflow::GetPaddingFromString(paddingStr, &padding).ok())
+    return op->emitError() << "fail to parse padding attributes\n";
+
+  ArrayRef<Attribute> dilations =
+      op->getAttrOfType<ArrayAttr>("dilations").getValue();
+  ArrayRef<Attribute> strides =
+      op->getAttrOfType<ArrayAttr>("strides").getValue();
+  ArrayRef<Attribute> explicit_paddings;
+  if (padding == tensorflow::Padding::EXPLICIT) {
+    // EXPLICIT padding mode and the associated attribute is attached to
+    // Conv2D.
+    explicit_paddings = op->getAttrOfType<ArrayAttr>("padding_list").getValue();
```

Comment:
nit: any documentation or references about the contracting attr names ?

Reply:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/mlir/hlo/include/mlir-hlo/Dialect/mhlo/IR/hlo_ops_base.td#L202